### PR TITLE
fix: avoid loading configs twice

### LINKED ISSUE #29: rocks-config loads configs twice
Hey there! Really liking rocks.nvim and its extensions, thanks!

I noticed my config for `nvim-lspconfig` was being loaded twice because I have a custom `set` function that detects if a global keymap is being set twice. By inspecting `rocks-config.nvim` code I noticed that right [here](https://github.com/nvim-neorocks/rocks-config.nvim/blob/51dfc03b80e8f246c9d9c295cd7f74f7ec26de30/lua/rocks-config/init.lua#L54) `plugin_heuristics` contained duplicate possibilities:

```
{ "rocks-git-nvim", "rocks-git", "rocks-git-nvim", "rocks-git-nvim-nvim" }
{ "tree-sitter-lua", "tree-sitter-lua", "tree-sitter-lua", "tree-sitter-lua-nvim" }
{ "tree-sitter-markdown", "tree-sitter-markdown", "tree-sitter-markdown", "tree-sitter-markdown-nvim" }
{ "coq_nvim", "coq_nvim", "coq_nvim", "coq_nvim-nvim" }
{ "emmet-vim", "emmet", "emmet-vim", "emmet-vim-nvim" }
{ "pytest-vim", "pytest", "pytest-vim", "pytest-vim-nvim" }
{ "vim-indent-object", "indent-object", "vim-indent-object", "vim-indent-object-nvim" }
{ "vim-over", "over", "vim-over", "vim-over-nvim" }
{ "vim-repeat", "repeat", "vim-repeat", "vim-repeat-nvim" }
{ "vim-surround", "surround", "vim-surround", "vim-surround-nvim" }
{ "zenburn-nvim", "zenburn", "zenburn-nvim", "zenburn-nvim-nvim" }
{ "lualine-nvim", "lualine", "lualine-nvim", "lualine-nvim-nvim" }
{ "none-ls-nvim", "none-ls", "none-ls-nvim", "none-ls-nvim-nvim" }
{ "nvim-lspconfig", "lspconfig", "nvim-lspconfig", "nvim-lspconfig-nvim" }
{ "nvim-treesitter", "treesitter", "nvim-treesitter", "nvim-treesitter-nvim" }
{ "plenary-nvim", "plenary", "plenary-nvim", "plenary-nvim-nvim" }
{ "rocks-config-nvim", "rocks-config", "rocks-config-nvim", "rocks-config-nvim-nvim" }
{ "rocks-nvim", "rocks", "rocks-nvim", "rocks-nvim-nvim" }
```

Notice that for every plugin the first and the third possibilities are the same. I'm not absolutely sure if that's intended (although I'm not sure if this makes much sense). But because of this the following `for` loop loads my configs twice, because that's the matching name I used for my files.

My "fix" here was adding a `break` to the `for` loop:

```lua
        for _, possible_match in ipairs(plugin_heuristics) do
            local mod_name = table.concat({ config.config.plugins_dir, possible_match }, ".")
            local ok = try_load_config(mod_name)
            found_custom_configuration = found_custom_configuration or ok
            if found_custom_configuration then
                break
            end
        end
```

And I guess I could open a Pull Request with that if that solution is reasonable. I think the `plugin_heuristics` table should be deduplicated, but even then, I believe once a plugin config is loaded the plugin should be considered configured and the search should stop anyway ¯\_(ツ)_/.

I'll get a PR ready here if my solution is reasonable.

### PR DESCRIPTION
**Pull request by tarcisioe**
Fixes possible points where the plugin can load the same configuration twice.

### CODE REVIEW

**Comment by tarcisioe** [comment]
This fixes the issue of loading configurations twice. There are three things to notice:

First, `create_plugin_heuristics` can create duplicate items. I added a `dedup` function to avoid even trying to load the same file twice. This change can be discarded since it is redundant with the rest, but I guess not trying the same thing twice can be less confusing if someone is stepping through the code.

Second, `try_load_config` checks for `package.loaded[mod_name]` to be true (actually not nil or false). However, since this code skips the `require` mechanism it wasn't doing something that `require` actually does: if the loaded module returns `nil` (which most configuration modules will, actually), `require` populates `package.loaded` with true. Since the code was using the result of `loader()`, `nil` was being added, therefore the check failed the next time. Modules can in fact return false as well, so I changed the check for an explicit `nil` check.

Third, the lookup for a configuration module didn't stop if a valid configuration was found, therefore it was possible for a user to have two configuration files for the same plugin. This is opinionated on my part, but I believe that shouldn't happen. In fact, it would even be possible to warn the user that a given plugin has two configuration files, and I would be open to implementing that as well if desired.

Closes #29.

File: lua/rocks-config/init.lua [hunk 1]
```diff
@@ -2,35 +2,77 @@ local constants = require("rocks-config.constants")
...
+            hash[v] = true
+        end
+    end
+
+    return res
+end
+
+---Creates plugin heuristics for a given plugin
 ---@param name string
+---@return table
```
Comment (mrcjkb) [inline lua/rocks-config/init.lua]: quibble:

```suggestion
---@return string[]
```

File: lua/rocks-config/init.lua [hunk 2]
```diff
@@ -2,35 +2,77 @@ local constants = require("rocks-config.constants")
 
 local rocks_config = {}
 
---- Creates plugin heuristics for a given plugin
+---Deduplicates a table that is being used as an array of strings
+---@param arr table
+---@return table
```
Comment (mrcjkb) [inline lua/rocks-config/init.lua]: quibble:

```suggestion
---Deduplicates a list of strings
---@param arr string[]
---@return string[]
```

**Review by mrcjkb — APPROVED** [review]
Thanks for the fix :smile: 

LGTM. Just some minor nitpicks.

**Comment by mrcjkb** [comment]
> Third, the lookup for a configuration module didn't stop if a valid configuration was found, therefore it was possible for a user to have two configuration files for the same plugin. This is opinionated on my part, but I believe that shouldn't happen. In fact, it would even be possible to warn the user that a given plugin has two configuration files, and I would be open to implementing that as well if desired.

@vhyrro what do you think about this? Personally, I'm fine with it.

**Comment by mrcjkb** [comment]
Looks like stylua is failing:

```
       > +++ b/lua/rocks-config/init.lua
       > @@ -42,7 +42,7 @@ local function try_load_like_require(searcher, mod_name)
       >      local loader = searcher(mod_name)
       >  
       >      if type(loader) ~= "function" then
       > -       return nil
       > +        return nil
       >      end
       >  
       >      local module = loader()
```

**Comment by vhyrro** [comment]
> > Third, the lookup for a configuration module didn't stop if a valid configuration was found, therefore it was possible for a user to have two configuration files for the same plugin. This is opinionated on my part, but I believe that shouldn't happen. In fact, it would even be possible to warn the user that a given plugin has two configuration files, and I would be open to implementing that as well if desired.
> 
> @vhyrro what do you think about this? Personally, I'm fine with it.

Yes I agree that having two different config files for the same plugin is in 90% of cases user error. We shouldn't allow that to happen, so I'm happy to warn when dupes are found :)

**Comment by tarcisioe** [comment]
Fixed typing and styles. @vhyrro I'll give the detection of duplicates with a warning a go on a separate PR because I'll need to give it a bit more thought about how I can detect it without actually attempting to load the second file.

**Pull request merged**